### ISSUE TITLE
Example code to create an invoice

### ISSUE THREAD

**Issue by githubjonny** [XeroAPI/xero-php-oauth2#191]
It would be a massive help if the script came with an example on submitting an invoice to drafts in the authorizedResource.php (or in a separate file) e.g. create an invoice with 3 example lineitems.

As it stands, its easy to get the starter php etc up and running and add customers and customer data but very hard (impossible for me at the moment see https://stackoverflow.com/questions/65232512/xero-api-struggling-to-add-a-basic-invoice) to get invoices setup. 

A simple example would be a massive help.

**Comment by SidneyAllen** [comment]
@githubjonny - Have you checked out our companion app for this SDK?
https://github.com/XeroAPI/xero-php-oauth2-app

It demonstrates many of the calls including createInvoice.
https://github.com/XeroAPI/xero-php-oauth2-app/blob/7a6e4240a24f192d25864ca8ef5ca1573a2a3094/example.php#L1442

**Comment by githubjonny** [comment]
Hi @SidneyAllen , thank you for your reply.

Yes, I have tried the code from example.php but i'm finding it very hard to combine the example given with the items I want to add to the invoice. Been trying to sus it out for 2 days but not really got anywhere.

In oauth1 there was an example on and xml insert which was perfect as it gave me the backbone to work with and in turn then expand on and get working. However, I am aware that oauth1 ends this month (?).

Basically, I just need to work out how set up very very basic submission of a draft invoice into xero using the php oauth (e.g. just 3 products).

**Comment by githubjonny** [comment]
Any chance you could provide an example of how to pass the minimum details of a product in` $lineitems = ""?`

**Comment by rodjsta** [comment]
Hi @githubjonny, you will need to create a LineItem object and use the setLineItems method on the Invoice object.

If you read the example that @SidneyAllen you should be able to work it out. You'll see that in the example, when they're setting the lineitem on the invoice object, they're calling a mock line item through a getLineItem method. 

I've attached a screenshot of that example method. 

![image](https://user-images.githubusercontent.com/32949856/101888426-d7456480-3be9-11eb-808a-1fb7c9af5e67.png)

**Comment by githubjonny** [comment]
Hello @rodjsta 
Thank you for your help.
I've been going over this for the past hour and am still struggling. I think the problem is (other than me!) that oauth1 had a working example (so it could be edited then built upon etc). Oauth2 whilst it does have an example, that example is mixed in with loads of other examples and for someone like myself (who has beginner-ish level of php understanding), its incredibly hard to work out which bits of code from which lines etc all go together etc. 

Oauth1 was perfect from my perspective, e.g. It allowed me to run it locally on WAMP and link to it from my website and the example code was basic enough to allow me to tweak it and to send over invoice details (no other features were needed).

I'll keep hammering away at it and thank you again for your help.

Do you know when Oauth1 is going to be stopped?

**Comment by wobinb** [comment]
The OAuth 2.0 SDK uses a fundamentally different approach, in that you create objects and pass these to the SDK. The SDK then does the work of creating the payload for you (now in JSON format).

To import an invoice you first create an Invoice object and populate it with the various data. This will typically include a Contact object and an array of LineItem objects.

Finally you place the Invoice object into an Array of it's own and pass this to the SDK.

There was a previous issue raised on a similar subject here:
https://github.com/XeroAPI/xero-php-oauth2/issues/122  

The example code here: https://github.com/XeroAPI/xero-php-oauth2#authorizedresourcephp does the same thing with a contact.

The various get and set functions are named in line with the field names in the documentation here: https://developer.xero.com/documentation/api/invoices
So for the status of the invoice you would use setStatus

**Comment by githubjonny** [comment]
Hello @wobinb 

Thank you for your help with this.

I've looked over all the above comments and am trying my best to get my head around it.

I've now got:

```
$lineitem = new XeroAPI\XeroPHP\Models\Accounting\LineItem;
$lineitem->setDescription('Sample Item')
		->setQuantity(1)
		->setUnitAmount(20)
		->setAccountCode("400");

$arr_lineitem = [];
array_push($arr_lineitem, $lineitem);



//var_dump($arr_lineitem);	
	
	
	

//[Invoices:Create]
$contact = new XeroAPI\XeroPHP\Models\Accounting\Contact;
$contact->setContactId($contactId);

$arr_invoices = [];	

$invoice_1 = new XeroAPI\XeroPHP\Models\Accounting\Invoice;
$invoice_1->setReference('Ref-')
	->setDueDate(new DateTime('2019-12-10'))
	->setContact($contact)
	->setLineItems($arr_lineitem)
	->setStatus(XeroAPI\XeroPHP\Models\Accounting\Invoice::STATUS_AUTHORISED)
	->setType(XeroAPI\XeroPHP\Models\Accounting\Invoice::TYPE_ACCPAY)
	->setLineAmountTypes(\XeroAPI\XeroPHP\Models\Accounting\LineAmountTypes::EXCLUSIVE);	
array_push($arr_invoices, $invoice_1);
	
$invoice_2 = new XeroAPI\XeroPHP\Models\Accounting\Invoice;
$invoice_2->setReference('Ref-')
	->setDueDate(new DateTime('2019-12-02'))
	->setContact($contact)
	->setLineItems($arr_lineitem)
	->setStatus(XeroAPI\XeroPHP\Models\Accounting\Invoice::STATUS_AUTHORISED)
	->setType(XeroAPI\XeroPHP\Models\Accounting\Invoice::TYPE_ACCPAY)
	->setLineAmountTypes(\XeroAPI\XeroPHP\Models\Accounting\LineAmountTypes::EXCLUSIVE);	
array_push($arr_invoices, $invoice_2);
			
$invoices = new XeroAPI\XeroPHP\Models\Accounting\Invoices;
$invoices->setInvoices($arr_invoices);

$result = $apiInstance->createInvoices($xeroTenantId,$invoices); 
//[/Invoices:Create]
```

which seems to be moving in the right direction.

When I check out https://developer.xero.com/app/apphistory/ I can see in the Request body 

`{"Invoices":[{"Type":"ACCPAY","Contact":{"HasAttachments":false,"HasValidationErrors":false},"LineItems":{"Description":"Sample Item","Quantity":1,"UnitAmount":20,"AccountCode":"400"},"DueDate":"2019-12-10T00:00:00+00:00","LineAmountTypes":"Exclusive","Reference":"Ref-","Status":"AUTHORISED","HasAttachments":false,"HasErrors":false},{"Type":"ACCPAY","Contact":{"HasAttachments":false,"HasValidationErrors":false},"LineItems":{"Description":"Sample Item","Quantity":1,"UnitAmount":20,"AccountCode":"400"},"DueDate":"2019-12-02T00:00:00+00:00","LineAmountTypes":"Exclusive","Reference":"Ref-","Status":"AUTHORISED","HasAttachments":false,"HasErrors":false}]}`

which does not have the invoice details. However, I am getting a status code of 500 so something still is incorrect.

**Comment by githubjonny** [comment]
Looking over the code, I think its contact ID (thats missing causing the problem??). Can this be in any format?

**Comment by githubjonny** [comment]
UPDATE : 

Changed line 
`$contact->setContactId($contactId);`
to an actual customer (changed to xxx for security)
`$contact->setContactId('xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx');`
this now gives a status code of 200 and shows in the request body:
`{"Invoices":[{"Type":"ACCPAY","Contact":{"ContactID":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","HasAttachments":false,"HasValidationErrors":false},"LineItems":[{"Description":"Sample Item","Quantity":1,"UnitAmount":20,"AccountCode":"400"}],"DueDate":"2019-12-02T00:00:00+00:00","LineAmountTypes":"Exclusive","Reference":"Ref-","Status":"AUTHORISED","HasAttachments":false,"HasErrors":false}]}`
However, nothing shows in xero under invoices

**Comment by githubjonny** [comment]
This is the current full code (minus security keys etc)
```

<?php
  ini_set('display_errors', 'On');
  require __DIR__ . '/vendor/autoload.php';
  require_once('storage.php');

  // Use this class to deserialize error caught
  use XeroAPI\XeroPHP\AccountingObjectSerializer;

  // Storage Classe uses sessions for storing token > extend to your DB of choice
  $storage = new StorageClass();
  $xeroTenantId = (string)$storage->getSession()['tenant_id'];

  if ($storage->getHasExpired()) {
    $provider = new \League\OAuth2\Client\Provider\GenericProvider([
      'clientId'                => 'REMOVED',
      'clientSecret'            => 'REMOVED',
      'redirectUri'             => 'https://REMOVED/callback.php',
      'urlAuthorize'            => 'https://login.xero.com/identity/connect/authorize',
      'urlAccessToken'          => 'https://identity.xero.com/connect/token',
      'urlResourceOwnerDetails' => 'https://api.xero.com/api.xro/2.0/Organisation'
    ]);

    $newAccessToken = $provider->getAccessToken('refresh_token', [
      'refresh_token' => $storage->getRefreshToken()
    ]);

    // Save my token, expiration and refresh token
    $storage->setToken(
        $newAccessToken->getToken(),
        $newAccessToken->getExpires(),
        $xeroTenantId,
        $newAccessToken->getRefreshToken(),
        $newAccessToken->getValues()["id_token"] );
  }

  $config = XeroAPI\XeroPHP\Configuration::getDefaultConfiguration()->setAccessToken( (string)$storage->getSession()['token'] );
  $apiInstance = new XeroAPI\XeroPHP\Api\AccountingApi(
      new GuzzleHttp\Client(),
      $config
  );
  

$lineitem = new XeroAPI\XeroPHP\Models\Accounting\LineItem;
$lineitem->setDescription('Sample Item')
		->setQuantity(1)
		->setUnitAmount(20)
		->setAccountCode("400");

$arr_lineitem = [];
array_push($arr_lineitem, $lineitem);

//var_dump($arr_lineitem);	
	
//[Invoices:Create]
$contact = new XeroAPI\XeroPHP\Models\Accounting\Contact;
$contact->setContactId('REMOVED aka xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx');

$arr_invoices = [];	

//$invoice_1 = new XeroAPI\XeroPHP\Models\Accounting\Invoice;
//$invoice_1->setReference('Ref-')
//	->setDueDate(new DateTime('2019-12-10'))
//	->setContact($contact)
//	->setLineItems($arr_lineitem)
//	->setStatus(XeroAPI\XeroPHP\Models\Accounting\Invoice::STATUS_AUTHORISED)
//	->setType(XeroAPI\XeroPHP\Models\Accounting\Invoice::TYPE_ACCPAY)
//	->setLineAmountTypes(\XeroAPI\XeroPHP\Models\Accounting\LineAmountTypes::EXCLUSIVE);	
//array_push($arr_invoices, $invoice_1);
	
$invoice_2 = new XeroAPI\XeroPHP\Models\Accounting\Invoice;
$invoice_2->setReference('Ref-')
	->setDueDate(new DateTime('2019-12-02'))
	->setContact($contact)
	->setLineItems($arr_lineitem)
	->setStatus(XeroAPI\XeroPHP\Models\Accounting\Invoice::STATUS_AUTHORISED)
	->setType(XeroAPI\XeroPHP\Models\Accounting\Invoice::TYPE_ACCPAY)
	->setLineAmountTypes(\XeroAPI\XeroPHP\Models\Accounting\LineAmountTypes::EXCLUSIVE);	
array_push($arr_invoices, $invoice_2);
			
$invoices = new XeroAPI\XeroPHP\Models\Accounting\Invoices;
$invoices->setInvoices($arr_invoices);


//[/Invoices:Create]


try {
    $result = $apiInstance->createInvoices($xeroTenantId,$invoices); 
    print_r($result);
} catch (Exception $e) {
    echo 'Exception when calling AccountingApi->createInvoices: ', $e->getMessage(), PHP_EOL;
}

?>
```

**Comment by githubjonny** [comment]
ok, just spotted it shows up as a "bill" in xero under the customer account instead of under invoices.

<img src="http://jizzdesigns.co.uk/example.JPG">

**Comment by githubjonny** [comment]
ACCREC!!!!!!

**Comment by githubjonny** [comment]
Pretty sure this isn't a bug and just me not fully understanding it but the code:

```

$setFirstName = "James";
$setLastName = "Price";
$setEmailAddress = "test@tests.com";

$contact = new XeroAPI\XeroPHP\Models\Accounting\Contact;
            $contact->setName($setFirstName . $setLastName)
                ->setFirstName($setFirstName)
                ->setLastName($setLastName)
                ->setEmailAddress($setEmailAddress);
                
            
            $arr_contacts = [];
            array_push($arr_contacts, $contact);
            
            $contacts = new XeroAPI\XeroPHP\Models\Accounting\Contacts;
            $contacts->setContacts($arr_contacts);

            $apiResponse = $apiInstance->createContacts($xeroTenantId,$contacts);
            $message = 'New Contact Name: ' . $apiResponse->getContacts()[0]->getName() .'<hr>' . $apiResponse->getContacts()[0]->getContactId() ;
```
will create a contact.

However, if I change everything but the name (so if someone else with the same name orders), it overwrites the original user with the new details.

Each of my customers on my website has a unique ID, is there a way to use this with xero instead of the xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx id?

**Comment by githubjonny** [comment]
Full code below:
```
// Define the vars
$setFirstName = "James";
$setLastName = "Price";
$setEmailAddress = "test@tests.com";
$setContactNumber = "13";

//[add a contact]
            $contact = new XeroAPI\XeroPHP\Models\Accounting\Contact;
            $contact->setName($setFirstName . $setLastName)
                ->setFirstName($setFirstName)
                ->setLastName($setLastName)
                ->setEmailAddress($setEmailAddress)
                ->setContactNumber($setContactNumber);
          
            $arr_contacts = [];
            array_push($arr_contacts, $contact);
            
            $contacts = new XeroAPI\XeroPHP\Models\Accounting\Contacts;
            $contacts->setContacts($arr_contacts);

            $apiResponse = $apiInstance->createContacts($xeroTenantId,$contacts);
            
            $message = 'New Contact Name: ' . $apiResponse->getContacts()[0]->getName() .'<hr>' . $apiResponse->getContacts()[0]->getContactId() ;
            
            echo "<hr>";
            echo $message;
            echo "<hr>";

//[/add a contact]
```

Can you see why this is not creating a new contact but instead overwriting the existing contact with the same name? Looking over example.php it would appear "$apiInstance->createContacts" would create a contact and "$apiInstance->updateOrCreateContacts" would be the update contact.

**Comment by rodjsta** [comment]
Hey @githubjonny nice work persisting through. Unfortunately Xero also requires Contact Name to be unique. You'll notice in the UI this is also the case. If you do some searching you'll also find many others who have complained before you. Xero have also said they're looking at making some changes to contacts in the future. I manage contacts  (and other objects) by saving the Xero guid on my end. 

In another application I've also appended an external ID number to the contact name if that name already exists in Xero. It seems like a very stupid limitation on Xero's behalf. I can't believe they're the size they are and still have this restriction.

**Comment by SidneyAllen** [comment]
Thanks @rodjsta for jumping in to help 👍
